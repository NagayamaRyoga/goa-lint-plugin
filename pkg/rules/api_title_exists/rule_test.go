package api_title_exists_test

import (
	"log"
	"os"
	"testing"

	"github.com/NagayamaRyoga/goalint/pkg/rules/api_title_exists"
	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/stretchr/testify/assert"
	"goa.design/goa/v3/dsl"
	"goa.design/goa/v3/eval"
	"goa.design/goa/v3/expr"
)

func TestRule(t *testing.T) {
	t.Parallel()

	var (
		apiWithTitle = dsl.API("calc1", func() {
			dsl.Title("Calculation API")
		})

		apiWithoutTitle = dsl.API("calc2", func() {
		})

		apiWithEmptyTitle = dsl.API("calc3", func() {
			dsl.Title("")
		})
	)

	// given
	testCases := []struct {
		description string
		api         *expr.APIExpr
		wantReports int
	}{
		{
			description: "success",
			api:         apiWithTitle,
			wantReports: 0,
		},
		{
			description: "failed/no title",
			api:         apiWithoutTitle,
			wantReports: 1,
		},
		{
			description: "failed/empty",
			api:         apiWithEmptyTitle,
			wantReports: 1,
		},
	}
	for _, tc := range testCases {
		t.Run(tc.description, func(t *testing.T) {
			t.Parallel()

			eval.Execute(tc.api.DSL(), tc.api)

			snaps := snaps.WithConfig(snaps.Dir("testdata"))

			logger := log.New(os.Stdout, "", 0)
			cfg := api_title_exists.NewConfig()
			rule := api_title_exists.NewRule(logger, cfg)

			// when
			got := rule.WalkMethodExpr(tc.api)
			assert.Len(t, got, tc.wantReports)
			snaps.MatchSnapshot(t, got.String())
		})
	}
}
