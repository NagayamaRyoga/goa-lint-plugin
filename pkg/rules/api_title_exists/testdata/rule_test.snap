
[TestRule/failed/empty - 1]
Found 1 errors and 0 warnings
[APITitleExists]: error in API("calc3"): API should have non-empty title

---

[TestRule/success - 1]

---

[TestRule/failed/no_title - 1]
Found 1 errors and 0 warnings
[APITitleExists]: error in API("calc2"): API should have non-empty title

---
